packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -355,6 +362,178 @@ class Actions extends InheritedWidget {
+}
+
+class _FocusableActionDetectorState extends State<FocusableActionDetector> {
+  @override
+  void initState() {
+    super.initState();
+    _updateHighlightMode(FocusManager.instance.highlightMode);
+    FocusManager.instance.addHighlightModeListener(_handleFocusHighlightModeChange);
+  }
+
+  @override
+  void dispose() {
+    FocusManager.instance.removeHighlightModeListener(_handleFocusHighlightModeChange);
+    super.dispose();
+  }
+
+  bool _canShowHighlight = false;
+  void _updateHighlightMode(FocusHighlightMode mode) {
+    final bool couldShowHighlight = _canShowHighlight;
+    switch (FocusManager.instance.highlightMode) {
+      case FocusHighlightMode.touch:
+        _canShowHighlight = false;
+        break;
+      case FocusHighlightMode.traditional:
+        _canShowHighlight = true;
+        break;
+    }
+    if  (couldShowHighlight != _canShowHighlight) {
+      _handleShowFocusHighlight();
+      _handleShowHoverHighlight();
+    }
+  }
+
+  /// Have to have this separate from the _updateHighlightMode because it gets
+  /// called in initState, where things aren't mounted yet.
+  void _handleFocusHighlightModeChange(FocusHighlightMode mode) {
+    if (!mounted) {
+      return;
+    }
+    _updateHighlightMode(mode);
+  }
+
+  bool _hovering = false;
+  void _handleMouseEnter(PointerEnterEvent event) {
+    assert(widget.onShowHoverHighlight != null);
+    if (!_hovering) {
+      // TODO(gspencergoog): remove scheduleMicrotask once MouseRegion event timing has changed.
+      scheduleMicrotask(() { setState(() { _hovering = true; _handleShowHoverHighlight(); }); });
+    }
+  }
+
+  void _handleMouseExit(PointerExitEvent event) {
+    assert(widget.onShowHoverHighlight != null);
+    if (_hovering) {
+      // TODO(gspencergoog): remove scheduleMicrotask once MouseRegion event timing has changed.
+      scheduleMicrotask(() { setState(() { _hovering = false; _handleShowHoverHighlight(); }); });
+    }
+  }
+
+  bool _focused = false;
+  void _handleFocusChange(bool focused) {
+    if (_focused != focused) {
+      setState(() {
+        _focused = focused;
+        _handleShowFocusHighlight();
+        widget.onFocusChange?.call(_focused);
+      });
+    }
+  }
+
+  void _handleShowHoverHighlight() {
+    widget.onShowHoverHighlight?.call(_hovering && widget.enabled && _canShowHighlight);
+  }
+
+  void _handleShowFocusHighlight() {
+    widget.onShowFocusHighlight?.call(_focused && widget.enabled && _canShowHighlight);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    Widget child = MouseRegion(
+      onEnter: _handleMouseEnter,
+      onExit: _handleMouseExit,
+      child: Focus(
+        focusNode: widget.focusNode,
+        autofocus: widget.autofocus,
+        canRequestFocus: widget.enabled,
+        onFocusChange: _handleFocusChange,
+        child: widget.child,
+      ),
+    );
+    if (widget.enabled && widget.actions != null && widget.actions.isNotEmpty) {
```

Comment:
Changing the shape of the widget tree (e.g. adding/removing a widget in the middle) is somewhat expansive. Is there a way to always include the Actions and Shortcuts widget, but have it "disabled" if its functionality is not required?